packages/php-wasm/universal/src/lib/php-process-manager.ts

```diff
@@ -0,0 +1,233 @@
+	 */
+	timeout?: number;
+	/**
+	 * The primary PHP instance that's never killed. This instance
+	 * contains the reference filesystem used by all other PHP instances.
+	 */
+	primaryPhp?: PHP;
+	/**
+	 * A factory function used for spawning new PHP instances.
+	 */
+	phpFactory?: PHPFactory<PHP>;
+}
+
+export interface SpawnedPHP<PHP extends BasePHP> {
+	php: PHP;
+	reap: () => void;
+}
+
+export class MaxPhpInstancesError extends Error {
+	constructor(limit: number) {
+		super(
+			`Requested more concurrent PHP instances than the limit (${limit}).`
+		);
+		this.name = 'MaxPhpInstancesError';
+	}
+}
+
+/**
+ * A PHP Process manager.
+ *
+ * Maintains:
+ * * A single "primary" PHP instance that's never killed – it contains the
+ *   reference filesystem used by all other PHP instances.
+ * * A pool of disposable PHP instances that are spawned to handle a single
+ *   request and reaped immediately after.
+ *
+ * When a new request comes in, PHPProcessManager yields the idle instance to handle it,
+ * and immediately starts initializing a new idle instance. In other words, for n concurrent
+ * requests, there are at most n+1 PHP instances running at the same time.
+ *
+ * A slight nuance is that the first idle instance is not initialized until the first
+ * concurrent request comes in. This is because many use-cases won't involve parallel
+ * requests and, for those, we can avoid eagerly spinning up a second PHP instance.
+ *
+ * This strategy is inspired by Cowboy, an Erlang HTTP server. Handling a single extra
+ * request can happen immediately, while handling multiple extra requests requires
+ * extra time to spin up a few PHP instances. This is a more resource-friendly tradeoff
+ * than keeping 5 idle instances at all times.
+ */
+export class PHPProcessManager<PHP extends BasePHP> implements Disposable {
+	private primaryPhp?: PHP;
+	private primaryIdle = true;
+	private nextInstance: Promise<SpawnedPHP<PHP>> | null = null;
+	/**
+	 * All spawned PHP instances, including the primary PHP instance.
+	 * Used for bookkeeping and reaping all instances on dispose.
+	 */
+	private allInstances: Promise<SpawnedPHP<PHP>>[] = [];
+	private phpFactory?: PHPFactory<PHP>;
+	private maxPhpInstances: number;
+	private semaphore: Semaphore;
+
+	constructor(options?: ProcessManagerOptions<PHP>) {
+		this.maxPhpInstances = options?.maxPhpInstances ?? 5;
+		this.phpFactory = options?.phpFactory;
+		this.primaryPhp = options?.primaryPhp;
+		this.semaphore = new Semaphore({
+			concurrency: this.maxPhpInstances,
+			/**
+			 * Wait up to 5 seconds for resources to become available
+			 * before assuming that all the PHP instances are deadlocked.
+			 */
+			timeout: options?.timeout || 5000,
+		});
+	}
+
+	/**
+	 * Get the primary PHP instance.
+	 *
+	 * If the primary PHP instance is not set, it will be spawned
+	 * using the provided phpFactory.
+	 *
+	 * @throws {Error} when called twice before the first call is resolved.
+	 */
+	async getPrimaryPhp() {
```

Comment:
Oftentimes we just want to call `mkdir` or `readFile` and it would be a waste to spin a new instance just to do that. Plus you can only do certain FS operations, like new mounts, on the primary instance.